Serving player - new feature

timonrey opened the pull request:
#### Description
With this new feature we will be able to see on the scoreboard which player has to serve. 

#### Hints for Review 
1. You can decide, which player has to start https://github.com/TimonRey/PingPongApp/pull/18/commits/0f00a0e0ea2c11c7ed6cb19a3687994eba533e50#diff-50781d54edd4e51a4c019a535ba764d9R127
2. You can change the serving player after two serves during the game
https://github.com/TimonRey/PingPongApp/pull/18/commits/0f00a0e0ea2c11c7ed6cb19a3687994eba533e50#diff-50781d54edd4e51a4c019a535ba764d9R131
3. In case of overtime, you can switch between the serving players after every point
https://github.com/TimonRey/PingPongApp/pull/18/commits/0f00a0e0ea2c11c7ed6cb19a3687994eba533e50#diff-50781d54edd4e51a4c019a535ba764d9R131

**Bug fixing**
1. The game did not reset after reaching two matchpoints ... After inserting the second if-else block into a new method, it worked again
3f28bcf40078e3353b2e32d4ca4eb0e47009dff9

2. After DoubleClick changed the score, the serving player was no longer correct. I've created a new method that determines the serving player after DoubleClick is pressed.
0d9f938bf4491196cb9ad7093e137ce16b5a6cf9

3. After one player starts a game, the other player should start the second game. With the new information, the program knows which player to start at the beginning of each game.
a8405c05a7c0eccf1f19cb29b4f66795c2c83985

**New UI is here!**
The HTML and CSS files have been replaced. Please check the new code:
1. newScoreboard.html: https://github.com/TimonRey/PingPongApp/pull/18/commits/665ce9a175f39ccf348dcae6bf4f5d6398d22708#diff-eb9ac053b2ee1b879bbfda566feffe12R1
2. newScoreboard.css: https://github.com/TimonRey/PingPongApp/pull/18/commits/665ce9a175f39ccf348dcae6bf4f5d6398d22708#diff-dca749570a8a33119cde1614c876dad3R1

Also please don't forget that we are using the template engine *thymeleaf* http://www.thymeleaf.org/

timonrey committed:
Fixing internet request error

The Flic buttons have always had an Internet request error, because the server wanted to load a template that does not exist

timonrey committed:
new method order + Serving feature logik

timonrey committed:
Implementing serving feature

timonrey committed:
Template text to see who is serving

timonrey committed:
save and change methods for serving

timonrey committed:
Delete endpoinds we don't need

emmenko commented on src/main/java/com/commercetools/pingpong/controller/GameController.java:
```diff
@@ -1,13 +1,12 @@
+        import org.springframework.http.HttpStatus;
+        import org.springframework.stereotype.Controller;
+        import org.springframework.ui.Model;
+        import org.springframework.web.bind.annotation.*;
```
Nitpick: something is wrong with the file formatting?

timonrey replied:
Yes, it's already fixed

timonrey replied:
https://github.com/TimonRey/PingPongApp/pull/18/commits/bbfaf6883a910e9f231fb277adea402282a086a1#diff-aa379526cf99272c6bd56552f614c180R3

emmenko commented on src/main/java/com/commercetools/pingpong/controller/GameController.java:
```diff
@@ -21,21 +20,40 @@
-    @RequestMapping(path = "/score")
-    public void updateScorePlusOne(@RequestParam(value="player", required = false) String player, Model model) {
+
+        public  String servingPlayer() {
```
Hmm, what is the purpose of this method? Are you sure you want to return a `String`?

timonrey replied:
I think so ... that was the easiest way to tell the template what to show on the scoreboard
But I am open to other suggestions

emmenko replied:
I see. Usually you would pass some data to the template page and let the page decide what to render/show.

Let's keep it like this for now, we will refactor this when we work on the frontend part.

emmenko commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -1,5 +1,7 @@
 package com.commercetools.pingpong.model;
 
+import com.sun.tools.javac.util.BasicDiagnosticFormatter;
```
What do you need this for?

timonrey replied:
I don't think we need this import anymore

emmenko commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -9,78 +11,101 @@ public Game(Player firstPlayer, Player secondPlayer) {
+    public void updateScoreOfPlayer(Player player, int actionType) {
+
+        if (actionType == 1 && !hasSomebodyWon()) {
+            player.add();
```
Unrelated to this PR, but I find this name a bit confusing. At first I thought a new player will be added 🙃

timonrey replied:
Yes, i will change it into **addPoint**

timonrey replied:
https://github.com/TimonRey/PingPongApp/pull/18/commits/bbfaf6883a910e9f231fb277adea402282a086a1#diff-50781d54edd4e51a4c019a535ba764d9R16

emmenko commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -9,78 +11,101 @@ public Game(Player firstPlayer, Player secondPlayer) {
+            resetServe();
+
+        } else if (this.playerTwo.getMatchScore() == 2) {
+            resetSetScores();
```
Nitpick: this function is called in every case, so you can simply define it once after the if-else conditions

timonrey replied:
I put both else ifs in a single condition and left them as else if in the block. But i don't understand why it should be defined after the if-else conditions

emmenko replied:
Just for simplicity:

```js
// before
if (...) {
  resetSetScores();
} else if (...) {
  resetSetScores();
} else if (...) {
  resetSetScores();
} else {
  resetSetScores();
}

// after
if () {
  ...
} else if () {
  ...
} else if () {
  ...
} else {
  ...
}
resetSetScores();

```

timonrey replied:
But in this case, the method would always reset the score, whether the conditions are correct or not

timonrey replied:
https://github.com/TimonRey/PingPongApp/pull/18/commits/bbfaf6883a910e9f231fb277adea402282a086a1#diff-50781d54edd4e51a4c019a535ba764d9R40

emmenko commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -9,78 +11,101 @@ public Game(Player firstPlayer, Player secondPlayer) {
+            resetSetScores();
+
+        }
+        if (this.playerOne.getMatchScore() == 2) {
```
Is this a different if-else block or was is supposed to be a `else-if` from the previous one?

emmenko replied:
Asking because if it's a different if-else block you are calling `resetSetScores` two times.

timonrey replied:
The second if/else block has a different purpose but it should be in a single block...you're right

timonrey replied:
https://github.com/TimonRey/PingPongApp/pull/18/commits/bbfaf6883a910e9f231fb277adea402282a086a1#diff-50781d54edd4e51a4c019a535ba764d9R40

timonrey replied:
The game did not reset after reaching two matchpoints ... After inserting the second if-else block into a new method, it worked again
3f28bcf40078e3353b2e32d4ca4eb0e47009dff9

emmenko commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -91,5 +116,36 @@ public Player getPlayerTwo() {
     }
 
-}
+    public boolean getPlayerOneServe() {return playerOne.amIServing();}
```
This function does not seem very useful

PhilippSpo replied:
True. One should just get the player and then call the `amIServing` method manually.

emmenko commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -91,5 +116,36 @@ public Player getPlayerTwo() {
+
+    public void decideWhoServes() {
+        if (areBothScoresZero()) {
+
```
Is it empty on purpose? If so you can simply return

```java
if (areBothScoresZero()) {
  return;
}
```

timonrey replied:
I didn't realize this is possible
Why does it work? The method is still void ..

emmenko replied:
Well `void` means that you don't return anything specific. So by saying `return;` you exit early from the method and do not return any specific value.

Hope that makes sense :)

timonrey replied:
Okay, I just did not expect that could work that way
Thanks for explaining 👍 😃

timonrey replied:
https://github.com/TimonRey/PingPongApp/pull/18/commits/bbfaf6883a910e9f231fb277adea402282a086a1#diff-50781d54edd4e51a4c019a535ba764d9R125

emmenko reviewed (changes requested):
Looks good overall, I left some comments to improve a couple of things.

timonrey committed:
formatting imports

timonrey committed:
fixing BUG 1

Game has not been reset after someone has reached two matchpoints

timonrey committed:
fixing BUG 2

After DoubleClick changed the score, the serving player was no longer correct

timonrey committed:
fixing BUG 3

After one player starts a game, the other player should start the second game. With the new information, the program knows which player to start at the beginning of each game

PhilippSpo commented on src/main/java/com/commercetools/pingpong/controller/GameController.java:
```diff
@@ -21,21 +20,40 @@
+    @ResponseBody
+    public String updateScoreReset(@RequestParam(value="player", required = false) String player, Model model) {
         gameService.updateScore(new Message(player, 3));
+        return "Score reseted";
```
https://www.verbix.com/webverbix/English/reset.html 🤓

timonrey replied:
hahaha yes, i saw that yesterday already 🤓 
4826e16e5763ae7cbe11b2983fe809e1721096c5

PhilippSpo commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -9,78 +10,112 @@ public Game(Player firstPlayer, Player secondPlayer) {
+            resetServe();
+            resetBeginningServe();
+        } else if (this.playerTwo.getMatchScore() == 2) {
+            resetSetScores();
```
We are calling the exact same methods in both cases. We can combine them in one condition.

timonrey replied:
of course 👍 
d5238189bea2597a7861e1fb2e59018c39ddeda6

PhilippSpo commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -91,5 +126,67 @@ public Player getPlayerTwo() {
+    public void setFirstServe(Player actingPlayer) {
+        actingPlayer.setBeginningServe();
+        actingPlayer.setServe();
+        }
```
This bracket is probably wrongly indented.

timonrey replied:
sorry ☺️ 
06b114eaf36bb234810d9ecf3eb6cd26186d2197

PhilippSpo commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -91,5 +126,67 @@ public Player getPlayerTwo() {
+        }
+
+
+    public void setBeginningServe() {
```
Both the game and the player have a `setBeginningServe` method? 🤔

timonrey replied:
Oh, that's right...
595f3e3389620dfe24617ce2375bb075f8b6d2b8

PhilippSpo commented on src/main/java/com/commercetools/pingpong/model/Player.java:
```diff
@@ -21,17 +23,30 @@ public Player(final String name) {
+    }
+
+    public void setBeginningServe() {
+        beginningServe = true;}
```
the bracket should go to the next line

timonrey replied:
Ok 👍 
0827390b5ff5d3237abe982a15b51c0c618275c2

PhilippSpo commented on src/main/java/com/commercetools/pingpong/model/Player.java:
```diff
@@ -21,17 +23,30 @@ public Player(final String name) {
+        serve = true;
+    }
+
+    public void changeServe() {
```
If `setServe` sets the serve to true, then setting the serve to `false` would make more sense via a `unsetServe` method.

timonrey replied:
Yes, that's better
76c872d8d0a425144640f13d4377bbdcd6e182c0

PhilippSpo commented on src/main/java/com/commercetools/pingpong/model/Player.java:
```diff
@@ -64,4 +79,7 @@ public int getMatchScore() {
 
+    public boolean amIServing() {return serve;}
+
+    public boolean didIServeAtBeginning() {return beginningServe;}
```
Why do we need to differentiate between beginning serve and serve? Is that really necessary?

timonrey replied:
Yeah, I think it is ... maybe you know why after seeing the fix:
a8405c05a7c0eccf1f19cb29b4f66795c2c83985

PhilippSpo replied:
True 👍 Thanks for clarifying.

PhilippSpo commented on src/main/java/com/commercetools/pingpong/service/impl/GameServiceImpl.java:
```diff
@@ -13,18 +13,19 @@
+        } else {
+            match.updateScoreOfPlayer(actingPlayer, message.getActionType());
             match.updateMatchScoreOfPlayers();
+            match.isGameOver();
```
From the name of the function one would think the function returns a Boolean indicating whether the game is over or not. But behind the sense this function has a lot of unexpected side-effects like resetting scores and serves. This should probably be better indicated.

timonrey replied:
Yes, i think that's better
7f478b8676bb69b6b7de7cba6f3d42fbbb92ffdb

PhilippSpo reviewed (changes requested):
Overall I am not sure if it makes sense to store the state of who is serving in the players. It can only be one of the two at any given time—either player one or player two. Now we have to remember that every time we set the serve to one player we have to unset the server on the other player. This could be avoided by having one variable that points to the player that is currently serving.

timonrey committed:
english noob fail

:cry:

emmenko reviewed (commented):
I also have a feeling that things are starting to get more complicated now.
At this point where things are working, we should probably take a step back and simplify some of those things.

Don't take it as a critic, it's actually very normal in software development to refactor things along the way, especially if the code gets more "complicated" as new features get implemented.

We can talk about it and do this in a separate PR.

timonrey committed:
simplifying condition isGameOver

timonrey committed:
indenting bracket

timonrey committed:
change name of method

timonrey committed:
indenting bracket (2)

timonrey committed:
create new changeServe methods

emmenko commented on src/main/java/com/commercetools/pingpong/model/Game.java:
```diff
@@ -9,78 +10,107 @@ public Game(Player firstPlayer, Player secondPlayer) {
 
-    public boolean hasOneOfThePlayersTwoPointsMore(Player playerWhoIsWinning, Player playerWhoIsLosing) {
-        return (playerWhoIsWinning.getSetScore() - playerWhoIsLosing.getSetScore() == 2);
+    public void isGameOver() {
```
We should probably split this into 2 functions:
* `public boolean canEndGame() { return this.playerOne.getMatchScore() == 2 || this.playerTwo.getMatchScore() == 2; }`
* `public void endGame() { ... }`

Then you can use those as:

```java
if (match.canEndGame()) {
  match.endGame();
}
```

timonrey replied:
Yes, actually there is a **canEndGame** already
7f478b8676bb69b6b7de7cba6f3d42fbbb92ffdb

emmenko reviewed (changes requested):
One last change.

Then we can focus on the refactoring and improvements.

timonrey committed:
formatting

timonrey committed:
simplifying isGameOver + formatting

emmenko commented on src/main/java/com/commercetools/pingpong/service/impl/GameServiceImpl.java:
```diff
@@ -7,29 +7,29 @@
         } else {
             match.updateScoreOfPlayer(actingPlayer, message.getActionType());
             match.updateMatchScoreOfPlayers();
-            match.isGameOver();
```
Actually I meant to do the `if (match.isMatchOver())` check here 😅

Like this it seems that you will always end the game, whenever you click the button.

timonrey replied:
oh ok 👍 
0f7ce81de1de624edee1ff4102681bc06ac727cb

timonrey committed:
set condition to gameOver into Gameserviceimpl

emmenko reviewed (approved):
👌

timonrey committed:
reset beginningServe for longPress + naming

timonrey committed:
new UI html + css

timonrey committed:
configure template engine

PhilippSpo commented on src/main/java/com/commercetools/pingpong/controller/GameController.java:
```diff
@@ -21,21 +22,62 @@
+    }
+
+    public String getIfLeftPlayerServes() {
+        if (gameService.getLeftPlayerServe()) {
```
We should move this logic into the template

PhilippSpo commented on src/main/java/com/commercetools/pingpong/controller/GameController.java:
```diff
@@ -21,21 +22,62 @@
+    }
+
+    public String getIfRightPlayerServes() {
+        if (gameService.getRightPlayerServe()) {
```
Same as above

timonrey committed:
testing font-family

timonrey committed:
Downloading font family
